Added user_roles migration, model, endpoint.

green-arrow opened the pull request:
Also added a tag to add to all endpoints that require interaction with AWS S3 that are ignored in tests by default.

To run all tests, including those which interface with S3, run `mix test --include s3`

Endpoint issue: #40
Contributes to #109 (exclude tests that rely on ENV vars)

begedin commented:
@JoshSmith 

I looked into support for base64 data with paperclip and it turns out, there is undocumented support, but only if the base64 string is in the standard data-uri format for images, meaning something like:

```
"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAfQAAAH0CAYAAADL1t+
KAAAKQWlDQ1BJQ0Mg\nUHJvZmlsZQAASA2dlndUU9kWh8+9N73QEiIgJfQaegkg0j
tIFQRRiUmAUAKG\nhCZ2RAVGFBEpVmRUwAFHhyJjRRQLg4Ji1wnyEFDGwVFERe
XdjGsJ7601896a\n/cdZ39nnt9fZZ+9917oAUPyCBMJ0WAGANKFYFO7rwVwSE8vE9
wIYEAEO..."
```

In that case, paperclip figures out the default adapter. Besides that, I don't think there's a way to embed the filename into the base64 data, so we need the original filename as a separate request parameter, but it seems you already accounted for that.

I modified `decode_image_data` so it should work by
* storing the original filename, with extension
* setting the right content type, based on the extension
* storing the image data

However, having the parameter contain the initial `"data:image/png;base64,"` part doesn't give much benefit API-side. I guess the client could be dumber in that case and just pass along the string, but depending on implementation, it might actually create a requirement on the client to add the string in the first place. 

Really, the only difference between our usual approach and the current one is that our `decode_image_data` is slightly simpler.

```Ruby
def decode_image_data
  return unless base_64_photo_data.present?
  data = Paperclip.io_adapters.for(base_64_photo_data)
  data.original_filename = self.filename
  self.image = data
end
```

I'm unable to write proper specs for it or test further, since I don't have the proper AWS credentials.

There's also the additional problem of something triggering a `HEAD` request to the resulting image url. I added VCR cassettes to the two specs that trigger it to get around it, but since I don't have the proper credentials, the returning values are "403 forbidden". Since we already mock our `Attachment#save` calls to just return `true`, I'm not sure why it's happening in the first place. Really, it should not be happening at all.

JoshSmith commented:
I still need to:

- [ ] Copypasta all this logic over for `CommentImage`
- [ ] Add validation for filenames

JoshSmith commented:
I'm not sure I like the way I'm testing the scenario getting called from the image worker.

JoshSmith commented:
Also would like to change the way I'm notifying pusher only once the image is decoded. That seems really tightly coupled.

JoshSmith commented:
The attachment validator can come in a separate PR. This is already ginormous.

JoshSmith commented:
Attachment validator is listed in #132 and policies in #131 #129

JoshSmith commented:
Otherwise this is good to review, and needs a really thorough look for anything I might have missed, or written poorly.

begedin commented on spec/factories/comment_image_factory.rb:
```diff
@@ -0,0 +1,17 @@
+    association :comment
+  end
+
+  trait :comment_with_s3_image do
```
I'd rename this to `with_s3_image`, so it's consistent with `post_image_factory`.

JoshSmith replied:
For some reason the trait won't work with the other trait that exists. Think it's because it's not wrapped in the factory itself? I don't know.

begedin commented:
Other than that one potential rename, I didn't notice any issues. Considering the size of the changeset, though, I can't guarantee it. Cautious :+1: from me.

green-arrow committed:
Added user_roles migration, model, endpoint.

Also added a tag to add to all endpoints that require secret env variables be ignored in tests by default.

To run all tests, including those which interface with secret env vars, run 'mix test --include requires_env'

Also added user_role_view_test, user_view_test, and preloaded roles
relationship on all user endpoints.

Endpoint issue: #40
Contributes to #109 (exclude tests that rely on ENV vars)

joshsmith commented on test/test_helper.exs:
```diff
@@ -2,6 +2,9 @@
 
+# By default, exclude S3 tests
+# To run S3 tests, run `mix test --include s3`
+ExUnit.configure exclude: [:s3]
```
I'm a little concerned about making this the default. I'd like to go the other way I think, and add the `exclude` command in the `README`.

We almost want to encourage people to get their own AWS account, too, so they can actually play around with some of that if they'd like to.

What do you think? It's scary to me to rely on a default that ignores some tests.

green-arrow replied:
So the reason I made this excluded by default is that these tests should be in the minority. The majority of the time, we won't be interacting with tests involving S3 / other secret env variables. Since it's likely most people won't mess with those tests, I think it makes sense to exclude by default and call it out in the README.

Also, as long as CI is set up to include those tests, we'll be sure to catch anything before it gets merged.

If we do want to include these by default, just let me know and I'll change it. Just wanted to give my two cents :)

joshsmith replied:
Okay, I just worry about missing this on CI. Already I have to remember to include it there in the config.

green-arrow replied:
Understandable. I can make it included by default. I'll make sure to point that out in the README as well.

joshsmith replied:
I mean, I can add to CI now if you'd like. I definitely see what you're saying. Maybe let's try it this way and see if it becomes an issue for folks later. Then we can fix.

joshsmith commented on priv/repo/migrations/20160822011624_create_user_role.exs:
```diff
@@ -0,0 +1,14 @@
+      timestamps()
+    end
+
+    create index(:user_roles, [:user_id, :role_id], unique: true, name: :index_projects_on_user_id_role_id)
```
Is the name strictly necessary or will it be auto-generated?

green-arrow replied:
I believe the name will be auto-generated. This was taken from the `user_skills` implementation. Do we prefer to use the auto-generated names?

joshsmith replied:
I prefer auto-generated, yes.

joshsmith commented on test/controllers/user_controller_test.exs:
```diff
@@ -126,6 +126,7 @@ defmodule CodeCorps.UserControllerTest do
       assert user.biography == "Just a test user"
     end
 
+    @tag :s3
```
Is there a more generalizable tag we could use here, like `requires_env` or something?

green-arrow replied:
Do we see there being a distinction between interacting with third party services / general / other secret env variables? I'm just wondering how granular we want to make it?

I guess we could always make it generically `requires_env` now then get more granular in the future if we ever need to.

joshsmith replied:
I think generic > granular now.

joshsmith commented:
I would add a `UserRoleViewTest` to this (you can see examples scattered throughout the codebase, including my open PR, which needs review _wink wink_).

I would also consider doing a `UserViewTest` and ensuring that the `has_many` roles through is present there.

Minor comments on this overall, and generally looks good. Still might want to re-review after changes, so maybe add a commit and don't squash so we can review only new changes.

green-arrow commented:
@JoshSmith for the `UserViewTest`, adding that relationship causes some other tests (like `UserController` to break because the relationship has not yet been loaded. I don't see an endpoint like `/users/:id/roles`, so do we want to force pre-fetching that relationship?

joshsmith commented:
@green-arrow yes, so there's no specific endpoint for it, but we'll preload the roles to avoid the N+1 hit while still having their ids present in the `relationships` for the view. It's necessary for the onboarding, for example, and elsewhere throughout the app.

joshsmith commented:
To be clear, though, we're not `include`-ing these. Ember will go out and hit them with coalesced requests to the `user_roles` endpoint.

green-arrow commented:
Gotcha, sounds good

green-arrow commented:
@JoshSmith - I believe this is ready for review again

joshsmith commented:
👍
